docs: Add a web section (#8374)
slint-ui/slint: docs/astro/src/content/docs/guide/platforms/web.mdx

---
<!-- Copyright © SixtyFPS GmbH <info@slint.dev> ; SPDX-License-Identifier: MIT -->
title: Web
description: Running Slint on the Web with Wasm
---

> While Slint is mainly designed for native applications (desktop, embedded, mobile),
> it can also run in a web browser using WebAssembly (Wasm).
> 
> :::caution[Caution]
> Currently, only the Rust language is supported when targeting the web, and the only renderer available is FemtoVG.
> :::

Reviewer: Here's a suggestion to phrase it more actively, instead of a passive observation of Slint:

(See section 6 of the style guide)

```suggestion
Slint applications run on desktop, embedded, and mobile platforms. Slint applications written with Rust can
also be cross-compiled to WebAssembly (Wasm) and run in the web browser.

:::note[Note]
Slint cross-compiled for Wasm uses the  <Link type="WinitBackend" label="Winit backend" /> with the FemtoVG renderer.
:::
```

Author: The point of the warning was to point out that rust is the only language.

Reviewer: Instead of that being a warning, I propose to make it part of the regular intro paragraph (second sentence). Caution in markdown notes are used to warn about potential negative consequences. That doesn't make sense here, I think.

Author: this just match what is in the tutorial.
I think it's not fair for the C++/JS people reading the docs if this doesn't emphasis that this will not work for them.

This renderer draws the UI into a `<canvas>` element on the page.
The rendering happens entirely within the canvas, so Slint does not rely on the browser's DOM or CSS engine.
This makes it possible to have a consistent look across platforms, but also introduces some limitations:

> * Slint handles its own font rendering, so it does not benefit from the browser's text rendering.
> * Accessibility features (such as screen readers) are not available.
> * The UI does not behave like a typical web application, since it doesn't use standard HTML elements.

Reviewer: ```suggestion
* Slint renders text directly, instead of benefitting from the browser's text rendering.
* Accessibility features (such as screen readers) are not available.
* The UI does not behave like a typical web application, since it doesn't use standard HTML elements.
```

Because of these trade-offs, running Slint in the browser is currently not recommended for building general-purpose web applications.
Instead, it is best suited for:

* Demos and examples that can run directly in the browser without requiring installation.
* Applications where the web is not the primary platform, but a consistent UI is still needed.
* Tools or dashboards where native-style rendering is more important than web integration.

> In the future, we plan to explore a renderer that uses the DOM directly, but this is not currently the focus.

Reviewer: We've had this a few times in the docs. I suggest we don't write about things in the future unless we're definitely committed to them.

```suggestion
```

> ## Building for WebAssembly

Reviewer: We could use the abbreviation here, after the introduction earlier.

```suggestion
## Building for Wasm
```

> To compile and run a Slint application in the browser, you will use
> [`wasm-bindgen`](https://rustwasm.github.io/docs/wasm-bindgen/) and [`wasm-pack`](https://rustwasm.github.io/wasm-pack/).

Reviewer: ```suggestion

The following pre-requisites are required:

 - [`wasm-bindgen`](https://rustwasm.github.io/docs/wasm-bindgen/)
 - [`wasm-pack`](https://rustwasm.github.io/wasm-pack/)
```

Author: is wasm-bindgen a pre-requisite?

Reviewer: It appears that `wasm-pack` installs it implicitly (not into `PATH` though), so it's indeed not necessary for users to install it. I'll edit the above suggestion.

> The tutorial has a chapter for a step-by-step walkthrough.

Reviewer: Since we're inside the same astro site, this really should be link to that chapter, IMO. Or at the very least to the tutorial.

(Also, unlike in German, for easier skimming it's great to mention the objective first ("For a walk-through"))

```suggestion
For a step-by-step walkthrough, check out the last chapter of the <Link type="quickstart" />.
```

Below is a summary of the main steps:

> In your `Cargo.toml`, set the crate type to `"cdylib"` and add `wasm-bindgen` as a dependency for the `wasm32` target:

Reviewer: In know right now we're only using `wasm32`, but there's `wasm64` on the horizon. I think we should use the correct new way, which is `target_family = "wasm"`. I've started doing this when I refactor, but for new content I think we should do that right away.

```toml
[lib]
#...
crate-type = ["cdylib"]

[target.'cfg(target_arch = "wasm32")'.dependencies]
wasm-bindgen = { version = "0.2" }
```

Use the `wasm-bindgen(start)` attribute to mark the application's entry point.
The UI is created and run as usual:

```rust
#[cfg(target_arch = "wasm32")]
use wasm_bindgen::prelude::*;

slint::include_modules!(); // or slint!(...)

#[cfg_attr(target_arch = "wasm32", wasm_bindgen(start))]
pub fn main() {
    // Usual application code
    let main_window = MainWindow::new().unwrap();
    main_window.run().unwrap();
}
```

Build the application using wasm-pack:

```bash
wasm-pack build --release --target web
```

This creates a `pkg/` directory with `.wasm` and `.js` files, including a JavaScript file named after your package.

> The wasm binary can be imported in a HTML file.
> Slint expects a `<canvas>` HTML element with `id = "canvas"`.

Reviewer: ```suggestion
Import the wasm binary in your HTML file.
Slint expects a `<canvas>` HTML element with `id = "canvas"`.
```

```html
<canvas id="canvas"></canvas>
<script type="module">
  import init from "./pkg/YOUR_APPLICATION.js";
  init();
</script>
```

Replace `YOUR_APPLICATION` with the name of your crate.

> Note that Wasm files can only be loaded in trusted contexts, so you may need to serve the files from a web server.

Reviewer: ```suggestion
:::note[Note]
Many web browser load `.wasm` files only in trusted contexts. If during development you observe the browser producing permission errors, then you may need to serve files through a web server, instead of the file system directly.
:::
```

> ## Clipboard Support in TextInput

Reviewer: ```suggestion
## Clipboard Support in `TextInput` and `LineEdit`
```

Author: TextInput is the only builtin widget.  LineEdit and TextEdit are widget. Feels stange to put TextInput and LineEdit but not TextEdit.  Maybe "`TextInput` and input widgets"?

Reviewer: Good point! How about `## Clipboard Support` and ... that's it? :)

> Slint's `TextInput` component supports copy and paste when running in the browser.

Reviewer: ```suggestion
Slint's text input widgets supports copy and paste when running in the browser.
```
